### DISCUSSION TITLE
.refine and .regex prevent .parse from returning .default value

### DISCUSSION THREAD

**Discussion by mneveroff** [colinhacks/zod#3696] [Q&A] [answered]
I'm using zod with react-hook-form and zod-i18n-map. I'm currently running up against an issue with a field that has a `.refine` statement, following the [documentation on the .default](https://zodjs.netlify.app/guide/schema-methods#default) usage.

There are a few issues I ran into and I was wondering if there's something that I'm doing wrong. Consider this schema:

```ts
const BaseSchema = z.object({
  type: z.enum('One', 'Two').default('One'),
  properties: z.object({
    name: z
      .string()
      .min(3)
      .max(64)
      // .regex(ALPHA_NUMERIC_REGEX, { message: 'errors.can-only-contain-alphanumeric-values' }),
      // Can't use regex https://github.com/aiji42/zod-i18n/issues/65
      .refine(
        (val) => {
          return val.match(ALPHA_NUMERIC_REGEX);
        },
        {
          params: { i18n: 'errors.can-only-contain-alphanumeric-values' },
        }
      )
      .default('New Object'),
  }),
});
```

With it, the call `BaseSchema.safeParse({ type: undefined, properties: {undefined}})` returns an error from val.match(), in my case it's `Object name can only contain letters (a-z, A-Z) and numbers (0-9)`. 

If I comment out the refine, the call to `BaseSchema.parse({ type: undefined, properties: {undefined}})` returns:

```json
{
  "type": "One",
  "properties": {
    "name": "New Object"
  }
```

as I would expect.

Same exact behaviour is observed when I use regex:

```
const BaseSchema = z.object({
  type: z.enum('One', 'Two').default('One'),
  properties: z.object({
    name: z
      .string()
      .min(3)
      .max(64)
      .regex(ALPHA_NUMERIC_REGEX, {
        message: i18n.t('zod:errors.alphanumeric_only_with_path', {
          path: i18n.t('form:properties.name'),
        }),
      })
      .default('New Object'),
  }),
});
```

Positioning `.default` before `.refine` doesn't change things, and the field is required so I'd rather not mark it as `.optional`. 
Is there something I'm missing? The reason I'm using refine instead of regex is [outlined in the Issue](https://github.com/aiji42/zod-i18n/issues/65) for zod-i18n-map and, to the best of my understanding, because zod's `.regex` doesn't allow passing parameters that would enable me to specify a custom error message.

**Comment by marilari88** [comment, marked as answer]
hi @MNeverOff, in the first schema, when name is undefined, the val inside the refine has 'New Object' string and it fails because of the whitespace not matching the regex. Take a look a this playground: [Zod playground with refine](https://zod-playground.vercel.app?appdata=N4IgzgxgFgpgtgQxALhALwHQHsBGArGCAFwApgAdAOwAJqiBPABxmWsxkoFc4SBtAciyUY-ADT8iAdyz8AugEoMAExgAzBJwA2pQcP7zRVWrUYAnLM1NEAljDCtMuAsTJHjtSgjgs2b97QwwIlNrSgBzEnk-fww4UJIAZiiafwDEAA8SADYAFmTUgNM1UJgSADcETWoAXgA%2BagoUguoiok5TGgrNWIQiaBIAemFJaidCIgHrfIKAX1EG6m8wMAQwn34IBEpKLCJqLuslXpgWmDX0-moZ6dTlNQ1tEn4AORgRgHl8cf1o68MU64gUQgLqcOwoXggRruMwWGBWWz2BbRTzeZAbawIGRUGZXIFQuhMdZCETzWGWGx2VjAK542TAsrwsDWIQoEAJDAAJg5AA4QDMgA )

Same behaviour we have with .regex() [Zod playground with regex](https://zod-playground.vercel.app?appdata=N4IgzgxgFgpgtgQxALhALwHQHsBGArGCAFwApgAdAOwAJqiBPABxmWsxkoFc4SBtAciyUY-ADT8iAdyz8AugEoMAExgAzBJwA2pQcP7zRVWrUYAnLM1NEAljDCtMuAsTJHjtSgjgs2b97QwwIlNrSgBzEnk-fww4UJIAZiiafwDEAA8SADYAFmTUgNMYMJhMgHphSWonQiIy61FqYGpvMDAEEtZ%2BCARKSiwiagA3BE1rJQQiGGoikvT%2BagBffILlNQ1tEn4AORgqgHl8Wv1o5cMU5ZBREBHNTjsUXhAKFJNzSxs7Vmboz29kbrWBAyKiLJZXZ50Jg%2BXQiRpmCwwKy2exNJbg2TXIZIsDWIQoEAJDAAJiJAA4QIsgA)

So wrapping up it seems like the default value is passed to regex and refine validations

Additionally, verify the enum syntax: it necessitates an array as an argument.

**Reply by mneveroff** [reply]
Oh shucks it's a nice catch on the regex not including the whitespace, no idea how that slipped my testing, thanks!
Noted on the enum, the actual implementation has an array (thanks TS), just my obfuscation for this question omitted it.
